packages/core/src/DataSchema.ts

```diff
@@ -0,0 +1,173 @@
+            253 * (this.config.FIELD_COUNT - this.config.SUM_FIELD_COUNT)
+
+        return schema.map((field, idx) => {
+            const { name, type, updateBy, ...extraFields } = field
+            const match = type.match(/^uint(\d+)$/)
+            if (match === null)
+                throw new Error(`Invalid type for field ${name}: "${type}"`)
+            if (Object.keys(extraFields).length !== 0) {
+                throw new Error(
+                    `Invalid fields included for field ${name}: [${Object.keys(
+                        extraFields
+                    )}]`
+                )
+            }
+
+            const duplicateEntry = schema.find(
+                (x, i) => x.name == field.name && idx != i
+            )
+            if (duplicateEntry) {
+                throw Error(
+                    `Schema includes a duplicate entry: "${duplicateEntry.name}"`
+                )
+            }
+
+            const bits = +match[1]
+            if (bits < 1 || bits > 253)
+                throw new Error(`Invalid uint size for field ${name}: ${bits}`)
+            if (updateBy === 'sum') {
+                if (
+                    Math.floor(sumOffset / 253) !==
+                    Math.floor((sumOffset + bits - 1) / 253)
+                ) {
+                    sumOffset += 253 - (sumOffset % 253)
+                }
+                if (sumOffset + bits > maxSumOffset) {
+                    throw new Error(
+                        `Invalid schema, field "${name}" exceeds available storage`
+                    )
+                }
+                const dataIndex = Math.floor(sumOffset / 253)
+                const offset = sumOffset % 253
+                sumOffset += bits
+                return { ...field, dataIndex, offset, bits }
+            } else if (updateBy === 'replace') {
+                if (bits !== 253 - this.config.REPL_NONCE_BITS)
+                    throw new Error(
+                        `Field must be ${
+                            253 - this.config.REPL_NONCE_BITS
+                        } bits`
+                    )
+                if (replOffset + bits > maxReplOffset) {
+                    throw new Error(
+                        `Invalid schema, field "${name}" exceeds available storage`
+                    )
+                }
+
+                const dataIndex =
+                    this.config.SUM_FIELD_COUNT + Math.floor(replOffset / 253)
+                const offset = replOffset % 253
+                replOffset += bits
+                return { ...field, dataIndex, offset, bits }
+            }
+            throw new Error(
+                `Invalid updateBy strategy for field ${name}: "${updateBy}"`
+            )
+        })
+    }
+
+    buildAttestation(change: { name: string; v: bigint }): Attestation {
+        const field: any = this.schema.find((f) => f.name === change.name)
+
+        if (field === undefined) {
+            throw new Error(`${change.name} not found`)
```

Comment:
Can be more explicit like `Field "${change.name}" not found`

Reply:
Change added,